### PR TITLE
Add error handling proposal documents

### PR DESCRIPTION
**Pull request by dbadura**
**Description**

Changes proposed in this pull request:

- add docs
**Related issue(s)**

#1325

### CODE REVIEW

File: docs/internal/proposals/error-handling/error-handling.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+This list can be split into 2 groups due to different handling:
+* internal errors
+* the rest of errors, which will be presented to the user
+
+In group of internal errors, are errors from:
+* external libraries
+* panics
+* postgreSQL, except errors like NotFound, NotUnique
+
+To deal with the errors we introduced custom errors, which contains error codes.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:25]:
```suggestion

```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+## Custom errors and theirs error codes:
+
+| Error type           | Error code  |                            Description                                                            |
+|----------------------|-------------|---------------------------------------------------------------------------------------------------|
+| InternalError        | 10          | Error which cannot be handled in director                                                         |
```
[the remark this answers is not in the document]
Comment (klaudiagrz) [reply docs/internal/proposals/error-handling/error-handling.md:31]: Please elaborate on this.

File: docs/internal/proposals/error-handling/error-handling.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+| NotFound             | 20          | Error indicate that given resource cannot be found and further processing is impossible           |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:32]:
```suggestion
| NotFound             | 20          | Indicates that a given resource cannot be found and further processing is impossible.           |
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+| NotUnique            | 21          | Error indicates that given resource is not unique                                                 |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:33]:
```suggestion
| NotUnique            | 21          | Indicates that a given resource is not unique.                                                 |
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+| InvalidData          | 22          | The input data is invalid, error description should be delivered in error message                 |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:34]:
```suggestion
| InvalidData          | 22          | Indicates that the input data is invalid with the reason described in the error message.                 | 
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+## Custom errors and theirs error codes:
+
+| Error type           | Error code  |                            Description                                                            |
+|----------------------|-------------|---------------------------------------------------------------------------------------------------|
+| InternalError        | 10          | Error which cannot be handled in director                                                         |
+| NotFound             | 20          | Error indicate that given resource cannot be found and further processing is impossible           |
+| NotUnique            | 21          | Error indicates that given resource is not unique                                                 |
+| InvalidData          | 22          | The input data is invalid, error description should be delivered in error message                 | 
+| InsufficientScopes   | 23          | Error which indicate that the client doesn't have sufficient permissions to execute the operation |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:35]:
```suggestion
| InsufficientScopes   | 23          | Indicates that the client doesn't have permission to execute the operation. |
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+| ConstraintViolation  | 24          | Error which indicate that this operation can't happen because referenced resource not exist       |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:36]:
```suggestion
| ConstraintViolation  | 24          | Indicates that this operation can't happen because the referenced resource does not exist.       |
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+| TenantIsRequired     | 25          | Tenant not found in request and is required to successful execute the request                     |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:37]:
```suggestion
| TenantIsRequired     | 25          | Indicates that a tenant is not found in the request, which is required to successfully execute the request.                     |
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+| TenantNotFound       | 26          | Internal Tenant not found in director                                                             |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:38]:
```suggestion
| TenantNotFound       | 26          | Indicates that the internal tenant is not found in the Director.                                                            |
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+NotFoundError can be triggered in mutation like `addPackage` to not existing application.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:40]:
```suggestion

```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+## Error processing flow:
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:42]:
```suggestion
## Error processing flow
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+![](error-handling.svg)
+
+Description of following steps
+* PostgreSQL Error Mapper - this component map postgreSQL errors to custom errors
+* Error presenter - this component search for custom error in error stack. 
+If such error is found, the presenter add `error_code` and `error` metadata to the GraphQL error in section `extensions`
+In case of internal errors, the whole error is logged and `internal server error` is sent to client.
+In case of errors which are not handled by custom error library, the error is returned without error code and error message is logged.
+
+## Proof of concept
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:53]:
```suggestion
## PoC
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+[Here](https://github.com/kyma-incubator/compass/pull/1366) is implemented PoC.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:54]:
```suggestion
See the [implemented PoC](https://github.com/kyma-incubator/compass/pull/1366) for more details.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+| TenantIsRequired     | 25          | Tenant not found in request and is required to successful execute the request                     |
+| TenantNotFound       | 26          | Internal Tenant not found in director                                                             |
+
+NotFoundError can be triggered in mutation like `addPackage` to not existing application.
+
+## Error processing flow:
+
+![](error-handling.svg)
+
+Description of following steps
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:46]:
```suggestion

```

Comment (klaudiagrz) [reply docs/internal/proposals/error-handling/error-handling.md:46]: Please describe the diagram in more detail. Maybe adding steps would be a good option? Refer to the [guidelines](https://github.com/kyma-project/community/blob/master/guidelines/content-guidelines/02-diagrams.md#examples) if needed.

File: docs/internal/proposals/error-handling/error-handling.md [hunk 15]
```diff
@@ -0,0 +1,50 @@
+# Error handling in Director GraphQL
+
+The errors can come from external libraries, triggered by the Director due to user wrong actions.
+We have to filter errors that are readable for both users and machines.
```
Comment (crabtree) [inline docs/internal/proposals/error-handling/error-handling.md:4]:
```suggestion
Errors can originate from different sources. This document describes our approach of handling them internally and presenting them in the readable form for the users.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+* InvalidData
+* InsufficientScopes
+* ConstraintViolation
+* TenantIsRequired
+* TenantNotFound
+
+This list can be split into two separate groups due to different error handling:
+* Internal errors that originate from external libraries, panics, PostgreSQL (except such errors as `NotFound`, `NotUnique`). 
+  The error is visible to the user as `InternalError`.
+* The rest of errors that are mostly caused by the user.
```
Comment (crabtree) [inline docs/internal/proposals/error-handling/error-handling.md:20]: I'd rephrase this paragraph, I do not see the need for presenting here the types as they are described in detail below.

```suggestion
Errors that are returned to the user can be split into two groups, based on the error handling approach. Errors that originate from external systems, libraries, panics, and directly not depend on the user input are handled as the internal one. Such errors are returned to the user as the `InternalError` type. Errors that directly depend on the user input are classified as separate types which in detail describes the nature of a problem.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 17]
```diff
@@ -0,0 +1,36 @@
+# Error handling in Director GraphQL
+
+Errors can originate from different sources. This document describes our approach of handling them internally and presenting them in the readable form for the users.
+
+## Errors displayed to the user
+Errors that are returned to the user can be split into two groups, based on the error handling approach. Errors that originate from external systems, libraries, panics, and directly not depend on the user input are handled as the internal one. Such errors are returned to the user as the `InternalError` type. Errors that directly depend on the user input are classified as separate types which in detail describes the nature of a problem.
```
Comment (tgorgol) [inline docs/internal/proposals/error-handling/error-handling.md:6]:
```suggestion
Errors that are returned to the user can be split into two groups, based on the error handling approach. Errors that originate from external systems, libraries, panics, and directly not depend on the user input are handled as the internal one. Such errors are returned to the user as the `InternalError` type. Errors that directly depend on the user input are classified as separate types which in detail describe the nature of a problem.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 18]
```diff
@@ -0,0 +1,36 @@
+# Error handling in Director GraphQL
+
+Errors can originate from different sources. This document describes our approach of handling them internally and presenting them in the readable form for the users.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:3]:
```suggestion
Errors can originate from different sources. This document describes our approach to handling them internally and presenting them in a readable form to the users.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+## Errors displayed to the user
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:5]:
```suggestion
## Errors displayed to the user

```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+|------------------------|---------------|-------------------------------------------------------------------------------------------------------------|
+| `InternalError`        | `10`          | Specifies any error that cannot be handled in the Director. Such situation is unique and serious.           |
+| `NotFound`             | `20`          | Indicates that a given resource cannot be found and further processing is impossible.                       |
+| `NotUnique`            | `21`          | Indicates that a given resource is not unique.                                                              |
+| `InvalidData`          | `22`          | Indicates that the input data is invalid with the reason described in the error message.                    | 
+| `InsufficientScopes`   | `23`          | Indicates that the client doesn't have permission to execute the operation.                                 |
+| `ConstraintViolation`  | `24`          | Indicates that this operation can't happen because the referenced resource does not exist.                  |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:20]:
```suggestion
| `ConstraintViolation`  | `24`          | Indicates that the operation cannot be executed because the referenced resource does not exist.                  |
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+| `TenantIsRequired`     | `25`          | Indicates that a tenant is not found in the request, which is required to successfully execute the request. |
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:21]:
```suggestion
| `TenantIsRequired`     | `25`          | Indicates that a tenant is not found in the request, which is required to successfully handle the request. |
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+| `TenantNotFound`       | `26`          | Indicates that the internal tenant is not found in the Director.                                            |
+
+## Error processing flow
+
+![](error-handling.svg)
+* Internal errors, External lib errors, PostgreSQL error, directives are sources of errors in the Director.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:27]:
```suggestion
1. Internal errors, external lib errors, PostgreSQL errors, and directives are sources of errors in the Director.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+* PostgreSQL Error Mapper - this component map PostgreSQL errors to custom errors
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:28]:
```suggestion
2. PostgreSQL Error Mapper maps PostgreSQL errors to custom errors.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+* Panic Handler - this component recovers all panics and transform them as custom errors
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:29]:
```suggestion
3. Panic Handler recovers all panics and transforms them into custom errors.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 25]
```diff
@@ -0,0 +1,36 @@
...
+* Error presenter - this component search for custom error in error stack.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:30]:
```suggestion
3. Error Presenter searches for custom errors in the error stack. 
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+| `TenantNotFound`       | `26`          | Indicates that the internal tenant is not found in the Director.                                            |
+
+## Error processing flow
+
+![](error-handling.svg)
+* Internal errors, External lib errors, PostgreSQL error, directives are sources of errors in the Director.
+* PostgreSQL Error Mapper - this component map PostgreSQL errors to custom errors
+* Panic Handler - this component recovers all panics and transform them as custom errors
+* Error presenter - this component search for custom error in error stack. 
+If such error is found, the presenter add `error_code` and `error` metadata to the GraphQL error in section `extensions`
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:31]:
```suggestion
If such an error is found, the Error Presenter adds `error_code` and `error` metadata to the GraphQL error in the **extensions** section.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+In case of internal errors, the whole error is logged and `internal server error` is sent to client.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:32]:
```suggestion
In case of internal errors, the error is logged and the `internal server error` message is sent to the client.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+In case of errors which are not handled by custom error library, the error is returned without error code and error message is logged.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:33]:
```suggestion
In case of errors that are not handled by custom error library, the error is returned without any error code and an error message is logged.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+
+## PoC
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:35]:
```suggestion
## PoC

```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+
+## Error processing flow
+
+![](error-handling.svg)
+1. Internal errors, external lib errors, PostgreSQL errors, and directives are sources of errors in the Director.
+2. PostgreSQL Error Mapper maps PostgreSQL errors to custom errors.
+3. Panic Handler recovers all panics and transforms them into custom errors.
+3. Error Presenter searches for custom errors in the error stack. 
+If such an error is found, the Error Presenter adds `error_code` and `error` metadata to the GraphQL error in the **extensions** section.
+In case of internal errors, the error is logged and the `internal server error` message is sent to the client.
```
Comment (dbadura) [inline docs/internal/proposals/error-handling/error-handling.md:33]:
```suggestion
In case of internal errors, the error is logged and the `internal server error` is sent to the client.
```

File: docs/internal/proposals/error-handling/error-handling.md [hunk 31]
```diff
@@ -0,0 +1,38 @@
+# Error handling in Director GraphQL
+
+Errors can originate from different sources. This document describes our approach to handling them internally and presenting them in a readable form to the users.
+
+## Errors displayed to the user
+
+Errors that are returned to the user can be split into two groups, based on the error handling approach. Errors that originate from external systems, libraries, panics, and directly not dependent on the user input are handled as the internal one. Such errors are returned to the user as the `InternalError` type. Errors that directly depend on the user input are classified as separate types.
```
Comment (klaudiagrz) [inline docs/internal/proposals/error-handling/error-handling.md:7]:
```suggestion
Errors that are returned to the user can be split into two groups, based on the error handling approach. Errors that originate from external systems, libraries, panics, and directly not dependent on the user input are handled as the internal ones. Such errors are returned to the user as the `InternalError` type. Errors that directly depend on the user input are classified as separate types.
```
